### DISCUSSION TITLE
How to call class functions inside cy.origin()

### DISCUSSION THREAD

**Discussion by parulbhagat-soprabanking** [cypress-io/cypress#27476] [Questions and Help]
We are trying to automate login to application after redirecting to Key Cloak and entering the login credentials on Key Cloak re-directed page.
We have classes for various functions like an Action class with all the object level functions like enter into a text field and we have separate classes to get the CSS properties of each of the page elements (basically following POM in cypress).
We want to use these class functions inside cy.origin(<key cloak re-directed URL>) , but are unable to do so.
How can we import a class inside a .js file and use the class functions inside cy.origin()?
Could you please provide us a way to sort out this issue ?
Thanks !

**Comment by Harvo09** [comment]
I had this. I could not pass a class POM into the args. 

What I did was create a .json file with my selectors for this test and passed that in the args.

mySelectors.json

```
{
  "selector1": "#myid",
  "selector2": ".myclass",
}  
```
Then Import the file into my test, then call my cy.origin code.

```
import mySelectors from "../elements/mySelectors.json";

 cy.origin(myotherURL,

            { args: { mySelectors } }, ({ mySelectors }) => {

                cy.get(mySelectors .selector1).click();
 });

```

**Comment by AskMarvinV** [comment]
Got this problem as well where we got POM pattern that we create class for variables and action. I'm stuck with this but i just decided not to use it for cy.origin. Here's an example:

UsernamePage.ts
```typescript
class UsernamePage {

   //Element Selector or Locator
   emailPlaceholderSelector: '[id="loginId"]';
   

   //Page Action 
   enterEmailId(emailid: string) {
      return cy.get(this.emailPlaceholderSelector).type(emailid)
   }
}

export default UsernamePage;

``` 

```typescript

function loginViaFusionAuth (username: string, password: string) {
    cy.visit('/')
    cy.origin('https://******/*/**', {args: {username, password }}, ({username, password}) => {
 
    cy.get('[data-test-id="TextInput-Login-Username"]').type(username);
    cy.get('#continue').click()
    cy.get('#password').type(password);
  
    cy.get('[data-test-id="Button-Login-Login"]').click()
  
  
  }
```

So when i replace this     cy.get('[data-test-id="TextInput-Login-Username"]').type(username);
with    enterEmailId(emailid: string) it's throwing error.

**Comment by DavidRS13** [comment]
I am facing the same issue trying to run class functions inside cy.origin() due to I have redirections as well and domain changes in a login. IT breaks POM approach. Is there any update about it?

**Comment by asliekm** [comment]
I couldn't call the functions within the origin, so I tried to call the functions from ./support/helper.js after defining them as arguments. But it still didn't work.

 cy.origin(
      'https://.....',
      { 
        args: { 
          commonLocators, 
          rLocators,
          helpers 
        } 
      },
      ({ commonLocators, rLocators, helpers }) =>

**Comment by malagamjaga** [comment]
I am also facing same issue where i am not able to use the selectors in the cy.origin()


I throws me this error

ReferenceError
_selectors is not defined

Variables must either be defined within the cy.origin() command or passed in using the args option.

Using require() or import() within the cy.origin() callback is not supported. Use Cypress.require() to include dependencies instead, but note that it currently requires enabling the experimentalOriginDependencies flag.[Learn more](https://on.cypress.io/origin)
login(username,password)
{

    cy.get(selectors.loginPage.loginButton).click();

        cy.origin(Cypress.env('loginUrl'), { args: {username, password } }, ({username, password }) => {
         // const selectors = Cypress.require('../../support/selectors');
          
          cy.get(selectors.loginPage.usernameInput)
            .should('be.enabled')
            .type(username);  // Use username passed from the test
    
          cy.get(selectors.loginPage.passwordInput)
            .should('be.enabled')
            .type(password);  // Use password passed from the test
    
          cy.get(selectors.loginPage.loginSubmit)
            .should('be.visible')
            .should('be.enabled')
            .click();
        });
